File: v4/rust/iprange-livedb/src/writer_core/publication.rs
//! Physical preparation and alternate-meta publication.

use crate::bootstrap::{Bootstrap, MetaSelection};
use crate::cancellation::CancellationToken;
use crate::contract::PAGE_SIZE;
use crate::draft_store::DraftStore;
use crate::error::{Error, Result};

use super::WriterCore;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub(crate) struct CommitAttempt {
    pub(crate) database_id: [u8; 16],
    pub(crate) transaction_id: u64,
    pub(crate) commit_nonce: [u8; 16],
}

#[derive(Debug)]
pub(crate) enum PublishOutcome {
    BeforePublication(Error),
    OutcomeUnknown(Error),
    Committed,
}

impl WriterCore {
    pub(crate) fn commit_attempt(&self) -> Result<CommitAttempt> {
        if self.workflow_input_open() {
            return Err(Error::WrongState("workflow input is not finished"));
        }
        let draft = self
            .draft
            .as_ref()
            .filter(|draft| draft.changed())
            .ok_or(Error::NoPendingTransaction)?;
        Ok(CommitAttempt {
            database_id: draft.meta.database_id,
            transaction_id: draft.meta.txn_id,
            commit_nonce: draft.meta.commit_nonce,
        })
    }

    pub(crate) fn prepare(&mut self, cancellation: &CancellationToken) -> Result<()> {
        let draft = self.draft.as_mut().ok_or(Error::NoPendingTransaction)?;
        let mut store = DraftStore::new(
            &mut self.mapping,
            self.base.meta.page_count,
            self.budget,
            draft,
        );
        store.prepare_with_checkpoint(&mut || cancellation.check())
    }

    pub(crate) fn require_draft_length(&self) -> Result<()> {
        let expected = self
            .draft
            .as_ref()
            .ok_or(Error::NoPendingTransaction)?
            .meta
            .page_count
            .checked_mul(PAGE_SIZE as u64)
            .ok_or(Error::ArithmeticOverflow("committed file length"))?;
        if self.mapping.file().metadata()?.len() < expected || self.mapping.len() < expected {
            return Err(Error::Corrupt("draft file length is inconsistent"));
        }
        Ok(())
    }

    pub(crate) fn publish(&mut self, cancellation: &CancellationToken) -> PublishOutcome {
        if let Err(error) = cancellation.check() {
            return PublishOutcome::BeforePublication(error);
        }
        crate::fault::crash("commit.before_private_sync");
        let committed_bytes = self
            .draft
            .as_ref()
            .expect("publication requires a draft")
            .meta
            .page_count
            * PAGE_SIZE as u64;
        let physical_bytes = match self.mapping.shrink_or_retain(committed_bytes) {
            Ok(physical_bytes) => physical_bytes,
            Err(error) => return PublishOutcome::BeforePublication(error),
        };
        let data_offset = (2 * PAGE_SIZE) as u64;
        if committed_bytes > data_offset {
            if let Err(error) = self
                .mapping
                .flush_range(data_offset, committed_bytes - data_offset)
            {
                return PublishOutcome::BeforePublication(error);
            }
        }
        if let Err(error) = self.mapping.sync_file() {
            return PublishOutcome::BeforePublication(error);
        }
        crate::fault::crash("commit.after_private_sync");
        if let Err(error) = cancellation.check() {
            return PublishOutcome::BeforePublication(error);
        }

        let meta = self
            .draft
            .as_ref()
            .expect("publication requires a draft")
            .meta;
        let target_page = 1 - self.base.selected_meta_page;
        let encoded = self
            .mapping
            .page_mut(u32::from(target_page), meta.page_count)
            .and_then(|page| meta.encode_mapped(page));
        if let Err(error) = encoded {
            return self.outcome_unknown(error);
        }
        crate::fault::crash("commit.after_meta_write");
        if let Err(error) = self
            .mapping
            .flush_page(u32::from(target_page), meta.page_count)
        {
            return self.outcome_unknown(error);
        }
        if let Err(error) = self.mapping.sync_file() {
            return self.outcome_unknown(error);
        }
        crate::fault::crash("commit.after_meta_sync");
        self.base = Bootstrap {
            meta,
            selection: MetaSelection::ProvenCurrent,
            selected_meta_page: target_page,
            committed_bytes,
            physical_bytes,
        };
        self.draft = None;
        self.unproved_tail_end = None;
        PublishOutcome::Committed
    }

    fn outcome_unknown(&mut self, error: Error) -> PublishOutcome {
        self.draft = None;
        self.unproved_tail_end = None;
        PublishOutcome::OutcomeUnknown(error)
    }
}

File: v4/go/internal/writer/publication.go
// Physical commit preparation and alternate-meta publication (Rust
// writer_core/publication.rs). Publish is the exact physical sequence:
// shrink-or-retain to the committed extent, flush the data pages, sync,
// write the alternate meta page into the mapping, flush it, sync again,
// then adopt the new generation as the proven-current base. The crash
// points of the Rust sequence are preserved test-only via internal/fault.

package writer

import (
	"github.com/firehol/iprange/v4/go/internal/bootstrap"
	"github.com/firehol/iprange/v4/go/internal/fault"
	"github.com/firehol/iprange/v4/go/internal/format"
)

// CommitAttempt is the identity of one prepared commit (Rust
// writer_core::CommitAttempt).
type CommitAttempt struct {
	DatabaseID    [16]byte
	TransactionID uint64
	CommitNonce   [16]byte
}

// PublishStatus classifies the outcome of one publish attempt (Rust
// PublishOutcome): the error arose before any publication step (the
// committed generation is untouched), after the outcome became unknown
// (the file may have advanced), or the commit landed.
type PublishStatus uint8

const (
	PublishCommitted PublishStatus = iota
	PublishBeforePublication
	PublishOutcomeUnknown
)

// PublishResult is one publish outcome (Rust PublishOutcome).
type PublishResult struct {
	Status PublishStatus
	Err    error
}

// StartDraft begins one COW draft over the committed generation with the
// caller's nonce (Rust workflow draft installation: Draft::new over
// base.meta). The all-zero nonce is refused exactly like Rust
// random::nonzero_128 refuses an all-zero draw: the commit nonce is the
// crash-recovery operation identity and a constant zero generation is a
// class Rust's API cannot produce. The public workflows draw the nonce
// from randomNonce when they arrive.
func (c *Core) StartDraft(nonce [16]byte) error {
	if err := c.requireHealthy(); err != nil {
		return err
	}
	if c.draft != nil {
		return &format.Error{Code: format.CodeWrongState, Detail: "a draft is already open"}
	}
	if nonce == [16]byte{} {
		return &format.Error{Code: format.CodeFormatInvalid, Detail: "commit nonce must be nonzero"}
	}
	draft, err := NewDraft(c.base.Meta, nonce)
	if err != nil {
		return err
	}
	c.draft = draft
	return nil
}

// BeginDraft starts one COW draft over the committed generation and
// draws the commit nonce (Rust WriterCore::begin_transaction: nonzero_128
// then Draft::new). The public workflows use this instead of accepting a
// caller nonce; the all-zero draw refusal of StartDraft stays in force
// (Rust random::nonzero_128 parity).
func (c *Core) BeginDraft() error {
	if err := c.requireHealthy(); err != nil {
		return err
	}
	nonce, err := randomNonce()
	if err != nil {
		return err
	}
	return c.StartDraft(nonce)
}

// Draft returns the open draft, or nil (Rust WriterCore::draft).
func (c *Core) Draft() *Draft { return c.draft }

// CommitAttempt names the pending commit while the workflow input is
// closed (Rust WriterCore::commit_attempt). The workflow-input-open gate
// is structurally closed: editor workflow states arrive with their public
// workflows. Exported for the public facade's commit orchestration.
func (c *Core) CommitAttempt() (CommitAttempt, error) {
	if err := c.requireHealthy(); err != nil {
		return CommitAttempt{}, err
	}
	if c.draft == nil || !c.draft.changed {
		return CommitAttempt{}, &format.Error{Code: format.CodeNoPendingTransaction, Detail: "no pending transaction"}
	}
	return CommitAttempt{
		DatabaseID:    c.draft.meta.DatabaseID,
		TransactionID: c.draft.meta.TxnID,
		CommitNonce:   c.draft.meta.CommitNonce,
	}, nil
}

// Prepare stages the draft for publication (Rust WriterCore::prepare):
// the draft store runs the full prepare-with-checkpoint sequence (private
// page release, bitmap shape, checksum sealing).
func (c *Core) Prepare(checkpoint func() error) error {
	if err := c.requireHealthy(); err != nil {
		return err
	}
	if err := fault.Fail("commit.prepare"); err != nil {
		return err
	}
	if c.draft == nil {
		return &format.Error{Code: format.CodeNoPendingTransaction, Detail: "no pending transaction"}
	}
	store := NewDraftStore(c.m, c.base.Meta.PageCount, c.budget, c.draft)
	return store.PrepareWithCheckpoint(checkpoint)
}

// RequireDraftLength verifies the locked file and the mapping both cover
// the draft's committed length (Rust WriterCore::require_draft_length:
// page_count * PAGE_SIZE, checked).
func (c *Core) RequireDraftLength() error {
	if c.draft == nil {
		return &format.Error{Code: format.CodeNoPendingTransaction, Detail: "no pending transaction"}
	}
	expected, err := checkedMul(c.draft.meta.PageCount, format.PageSize, "committed file length")
	if err != nil {
		return err
	}
	fileLen, err := c.m.FileSize()
	if err != nil {
		return err
	}
	if fileLen < expected || c.m.Size() < expected {
		return corrupt("draft file length is inconsistent")
	}
	return nil
}

// Publish commits the prepared draft by publishing the alternate meta
// page (Rust WriterCore::publish). On success the base becomes the new
// generation with ProvenCurrent selection and the draft is cleared; a
// failure before the meta write leaves the committed generation
// untouched, a failure after it reports OutcomeUnknown with the draft
// abandoned (no further use of this Core is safe).
func (c *Core) Publish(checkpoint func() error) PublishResult {
	if err := c.requireHealthy(); err != nil {
		return PublishResult{Status: PublishBeforePublication, Err: err}
	}
	if checkpoint != nil {
		if err := checkpoint(); err != nil {
			return PublishResult{Status: PublishBeforePublication, Err: err}
		}
	}
	fault.Crash("commit.before_private_sync")
	if c.draft == nil {
		return PublishResult{Status: PublishBeforePublication, Err: &format.Error{Code: format.CodeNoPendingTransaction, Detail: "no pending transaction"}}
	}
	committedBytes := c.draft.meta.PageCount * format.PageSize
	if err := c.m.Shrink(committedBytes); err != nil {
		return PublishResult{Status: PublishBeforePublication, Err: err}
	}
	dataOffset := uint64(2 * format.PageSize)
	if committedBytes > dataOffset {
		if err := c.m.FlushRange(dataOffset, committedBytes-dataOffset); err != nil {
			return PublishResult{Status: PublishBeforePublication, Err: err}
		}
	}
	if err := c.m.SyncFile(); err != nil {
		return PublishResult{Status: PublishBeforePublication, Err: err}
	}
	fault.Crash("commit.after_private_sync")
	if checkpoint != nil {
		if err := checkpoint(); err != nil {
			return PublishResult{Status: PublishBeforePublication, Err: err}
		}
	}

	meta := c.draft.meta
	target := uint8(1) - c.base.SelectedMetaPage
	page, err := c.m.Page(uint32(target))
	if err != nil {
		return c.outcomeUnknown(err)
	}
	if err := meta.EncodeMapped(page); err != nil {
		return c.outcomeUnknown(err)
	}
	fault.Crash("commit.after_meta_write")
	if err := fault.Fail("commit.after_meta_write"); err != nil {
		return c.outcomeUnknown(err)
	}
	if err := c.m.FlushPage(uint32(target)); err != nil {
		return c.outcomeUnknown(err)
	}
	if err := c.m.SyncFile(); err != nil {
		return c.outcomeUnknown(err)
	}
	fault.Crash("commit.after_meta_sync")
	physical, err := c.m.FileSize()
	if err != nil {
		return c.outcomeUnknown(err)
	}
	c.base = bootstrap.Result{
		Meta:             meta,
		Selection:        bootstrap.SelectionProvenCurrent,
		SelectedMetaPage: target,
		CommittedBytes:   committedBytes,
		PhysicalBytes:    physical,
	}
	c.draft = nil
	c.unprovedTailEnd = nil
	return PublishResult{Status: PublishCommitted}
}

// outcomeUnknown abandons the draft after a publication step whose effect
// on the file is unknown (Rust WriterCore::outcome_unknown) and brands
// the core unusable: the durability state of the file is unknown, so
// every mutating entry point fails closed until Close.
func (c *Core) outcomeUnknown(err error) PublishResult {
	c.draft = nil
	c.unprovedTailEnd = nil
	c.unresolved = err
	return PublishResult{Status: PublishOutcomeUnknown, Err: err}
}

// checkedMul multiplies with the Rust ArithmeticOverflow class.
func checkedMul(a, b uint64, what string) (uint64, error) {
	if a != 0 && b > ^uint64(0)/a {
		return 0, overflow(what)
	}
	return a * b, nil
}
